src/model.cpp

```diff
@@ -0,0 +1,250 @@
+
+            auto group = domain_.get_group(i);
+            std::vector<util::unique_any> cells(group.end-group.begin);
+
+            for (auto gid: util::make_span(group.begin, group.end)) {
+                auto i = gid-group.begin;
+                cells[i] = rec.get_cell(gid);
+            }
+
+            cell_groups_[i] = make_cell_group(
+                    group.kind, group.begin, std::move(cells), domain_.backend());
+            PL(2);
+        });
+
+    // store probes
+    for (const auto& c: cell_groups_) {
+        util::append(probes_, c->probes());
+    }
+
+    // generate the network connections
+    for (cell_gid_type i: util::make_span(domain_.cell_begin(), domain_.cell_end())) {
+        for (const auto& cc: rec.connections_on(i)) {
+            connection conn{cc.source, cc.dest, cc.weight, cc.delay};
+            communicator_.add_connection(conn);
+        }
+    }
+    communicator_.construct();
+
+    // Allocate an empty queue buffer for each cell group
+    // These must be set initially to ensure that a queue is available for each
+    // cell group for the first time step.
+    current_events().resize(num_groups());
+    future_events().resize(num_groups());
+}
+
+void model::reset() {
+    t_ = 0.;
+    for (auto& group: cell_groups_) {
+        group->reset();
+    }
+
+    communicator_.reset();
+
+    for(auto& q : current_events()) {
+        q.clear();
+    }
+    for(auto& q : future_events()) {
+        q.clear();
+    }
+
+    current_spikes().clear();
+    previous_spikes().clear();
+
+    util::profilers_restart();
+}
+
+time_type model::run(time_type tfinal, time_type dt) {
+    // Calculate the size of the largest possible time integration interval
+    // before communication of spikes is required.
+    // If spike exchange and cell update are serialized, this is the
+    // minimum delay of the network, however we use half this period
+    // to overlap communication and computation.
+    time_type t_interval = communicator_.min_delay()/2;
+
+    time_type tuntil;
+
+    // task that updates cell state in parallel.
+    auto update_cells = [&] () {
+        threading::parallel_for::apply(
+            0u, cell_groups_.size(),
+             [&](unsigned i) {
+                auto &group = cell_groups_[i];
+
+                PE("stepping","events");
+                group->enqueue_events(current_events()[i]);
+                PL();
+
+                group->advance(tuntil, dt);
+
+                PE("events");
+                current_spikes().insert(group->spikes());
+                group->clear_spikes();
+                PL(2);
+            });
+    };
+
+    // task that performs spike exchange with the spikes generated in
+    // the previous integration period, generating the postsynaptic
+    // events that must be delivered at the start of the next
+    // integration period at the latest.
+    auto exchange = [&] () {
+        PE("stepping", "communication");
```

Comment:
Agreed. These heavy duty poisson generators that you are talking about should not be implemented as a cell type in the long run (for now they can).

Sam and I feel that they should be implemented as a synapse type in NMODL. Because the spike train that they generate is consumed by only one synapse, so there is no need to pass the spike information through the spike-event system.